src/hotspot/share/gc/serial/markSweep.cpp

```diff
@@ -172,15 +172,30 @@ void MarkSweep::mark_object(oop obj) {
     _string_dedup_requests->add(obj);
   }
 
-  // some marks may contain information we need to preserve so we store them away
-  // and overwrite the mark.  We'll restore it at the end of markSweep.
-  markWord mark = obj->mark();
-  obj->set_mark(markWord::prototype().set_marked());
+  if (UseCompactObjectHeaders) {
+    // Do the transform while we still have the header intact,
+    // which might include important class information.
+    ContinuationGCSupport::transform_stack_chunk(obj);
+
+    // some marks may contain information we need to preserve so we store them away
+    // and overwrite the mark.  We'll restore it at the end of markSweep.
+    markWord mark = obj->mark();
+    obj->set_mark(obj->prototype_mark().set_marked());
+
+    if (obj->mark_must_be_preserved(mark)) {
+      preserve_mark(obj, mark);
+    }
+  } else {
+    // some marks may contain information we need to preserve so we store them away
+    // and overwrite the mark.  We'll restore it at the end of markSweep.
+    markWord mark = obj->mark();
+    obj->set_mark(markWord::prototype().set_marked());
 
-  ContinuationGCSupport::transform_stack_chunk(obj);
+    ContinuationGCSupport::transform_stack_chunk(obj);
 
-  if (obj->mark_must_be_preserved(mark)) {
-    preserve_mark(obj, mark);
+    if (obj->mark_must_be_preserved(mark)) {
```

Comment:
this if block could be put out of outer if?

Reply:
It needs markWord mark;